compiler/test/dotty/tools/dotc/printing/PrintingTest.scala

```diff
@@ -0,0 +1,84 @@
+package dotty
+package tools
+package dotc
+
+import vulpix.TestConfiguration
+import vulpix.TestConfiguration
+import reporting.TestReporter
+
+import java.io._
+import java.nio.file.{Path => JPath}
+import java.lang.System.{lineSeparator => EOL}
+
+import interfaces.Diagnostic.INFO
+import dotty.tools.io.Directory
+
+import scala.io.Source
+import org.junit.Test
+
+class PrintingTest {
+  val testsDir = "tests/printing"
+  val options = List("-Xprint:frontend", "-color:never", "-classpath", TestConfiguration.basicClasspath)
+
+  private def fileContent(filePath: String): List[String] =
+    if (new File(filePath).exists)
+      Source.fromFile(filePath, "UTF-8").getLines().toList
+    else Nil
+
+
+  private def compileFile(path: JPath): Boolean = {
+    val baseFilePath  = path.toAbsolutePath.toString.stripSuffix(".scala")
+    val outFilePath   = baseFilePath + ".out"
+    val checkFilePath = baseFilePath + ".check"
+    val ps = new PrintStream(new File(outFilePath))
+    val reporter = TestReporter.reporter(ps, INFO)
+
+    try {
+      Main.process((path.toString::options).toArray, reporter, null)
+    } catch {
+      case e: Throwable =>
+        println(s"Compile $path exception:")
+        e.printStackTrace()
+    }
+    finally ps.close()
+
+    val actual   = fileContent(outFilePath)
+    val expected = fileContent(checkFilePath)
+
+    val success =
+      actual.length == expected.length &&
+      (actual, expected).zipped.forall(_ == _)
+
+    success || {
+      println(
```

Comment:
This logic is already present in ParallelTesting. It would be nice if we find a way to share it.

Reply:
Good idea. What about we move the string construction logic to a new file `object FileDiff`?

Initially, I thought to extend ParallelTesting, but it seems there things are already complicated.

Reply:
Yes, I think factoring out string construction is a good solution